pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -299,47 +296,97 @@ func (s *Service) sendMsgEvents(msg *service.StateMsg) {
 func (s *Service) startInternalListener() {
 	go func() {
 		for msg := range s.callbackChannel {
-			// TODO handle error in callback - https://github.com/hyperledger/aries-framework-go/issues/242
-			if err := s.process(msg); err != nil {
-				// TODO handle error
-				logger.Errorf(err.Error())
+			if msg.Err != nil {
+				// TODO https://github.com/hyperledger/aries-framework-go/issues/438 Cleanup/Update data on Stop Event Action
+				logger.Errorf("client action event processing failed - msgID:%s error:%s", msg.ID, msg.Err)
+				err := s.processCallbackFailure(msg.ID, msg.Err)
+				if err != nil {
+					logger.Errorf("process callback failure: %s", err)
+				}
+				continue
+			}
+
+			// TODO https://github.com/hyperledger/aries-framework-go/issues/242 - retry logic
+
+			if err := s.process(msg.ID); err != nil {
+				procErr := s.processCallbackFailure(msg.ID, err)
+				if procErr != nil {
+					logger.Errorf("process callback failure: %s", procErr)
+				}
 			}
 		}
 	}()
 }
 
+func (s *Service) processCallbackFailure(id string, processErr error) error {
+	// get the transient data
+	data, err := s.getTransientEventData(id)
+	if err != nil {
+		return fmt.Errorf("unable to fetch the event transient data: %w", err)
+	}
+
+	err = s.processFailure(data.ThreadID, data.Msg, processErr)
+	if err != nil {
+		return fmt.Errorf("process failure: %w", err)
+	}
+
+	return nil
+}
+
+// processFailure updates the state to abandoned and trigger failure event.
+func (s *Service) processFailure(thid string, msg *service.DIDCommMsg, processErr error) error {
+	// update the state to abandoned
+	err := s.update(thid, &abandoned{})
+	if err != nil {
+		return fmt.Errorf("unable to update the state to abandoned: %w", err)
+	}
+
+	// send the message event
+	s.sendMsgEvents(&service.StateMsg{
+		Type:       service.PostState,
+		Msg:        msg,
+		StateID:    (&abandoned{}).Name(),
```

Comment:
Use constant instead of `&abandoned{}).Name()`

Reply:
I feel better to use the method as it'll take care of change in state name. If we point to a constant, then we might need to refactor (provided referring to 2 different constants).

Reply:
I also think a constant is better here.

Reply:
[Done](https://github.com/hyperledger/aries-framework-go/compare/807adcf62e0ac91e48d81a5b15a62d91e073bc78..0360003cb9f51624a1493097c2398be2113925f0#diff-f761c91e753eba1c42a68076df06ce12R345).